src/jobservice/job/impl/scandataexport/scan_data_export.go

```diff
@@ -0,0 +1,345 @@
+	digestCalculator      export.ArtifactDigestCalculator
+	filterProcessor       export.FilterProcessor
+	vulnDataSelector      export.VulnerabilityDataSelector
+	projectMgr            project.Manager
+	sysArtifactMgr        systemartifact.Manager
+}
+
+func (sde *ScanDataExport) MaxFails() uint {
+	return 1
+}
+
+// MaxCurrency of the job. Unlike the WorkerPool concurrency, it controls the limit on the number jobs of that type
+// that can be active at one time by within a single redis instance.
+// The default value is 0, which means "no limit on job concurrency".
+func (sde *ScanDataExport) MaxCurrency() uint {
+	return 0
+}
+
+// ShouldRetry tells worker if retry the failed job when the fails is
+// still less that the number declared by the method 'MaxFails'.
+//
+// Returns:
+//  true for retry and false for none-retry
+func (sde *ScanDataExport) ShouldRetry() bool {
+	return true
+}
+
+// Validate Indicate whether the parameters of job are valid.
+// Return:
+// error if parameters are not valid. NOTES: If no parameters needed, directly return nil.
+func (sde *ScanDataExport) Validate(params job.Parameters) error {
+	return nil
+}
+
+// Run the business logic here.
+// The related arguments will be injected by the workerpool.
+//
+// ctx Context                   : Job execution context.
+// params map[string]interface{} : parameters with key-pair style for the job execution.
+//
+// Returns:
+//  error if failed to run. NOTES: If job is stopped or cancelled, a specified error should be returned
+//
+func (sde *ScanDataExport) Run(ctx job.Context, params job.Parameters) error {
+	if _, ok := params[export.JobModeKey]; !ok {
+		return errors.Errorf("No mode specified for scan data export execution.")
+	}
+
+	mode := params[export.JobModeKey].(string)
+	logger.Infof("Scan data export job started in mode : %v", mode)
+	sde.init()
+	fileName := fmt.Sprintf("%s/scandata_export_%v.csv", sde.scanDataExportDirPath, params["JobId"])
+
+	// ensure that CSV files are cleared post the completion of the Run.
+	defer sde.cleanupCsvFile(fileName, params)
+	err := sde.writeCsvFile(ctx, params, fileName)
+	if err != nil {
+		logger.Errorf("Error when writing data to CSV: %v", err)
+		return err
+	}
+
+	hash, err := sde.calculateFileHash(fileName)
+	if err != nil {
+		logger.Errorf("Error when calculating checksum for generated file: %v", err)
+		return err
+	}
+	logger.Infof("Export Job Id = %v, FileName = %s, Hash = %v", params["JobId"], fileName, hash)
+
+	csvFile, err := os.OpenFile(fileName, os.O_RDONLY, os.ModePerm)
+	if err != nil {
+		logger.Errorf(
+			"Export Job Id = %v. Error when moving report file %s to persistent storage: %v", params["JobId"], fileName, err)
+		return err
+	}
+	baseFileName := filepath.Base(fileName)
+	repositoryName := strings.TrimSuffix(baseFileName, filepath.Ext(baseFileName))
+	logger.Infof("Creating repository for CSV file with blob : %s", repositoryName)
+	stat, err := os.Stat(fileName)
```

Comment:
what about if the size is 0?

Reply:
Currently, an empty file gets generated. However in my opinion this is ok. An empty CSV indicates to the user that his/her search criteria did not yield any results. This is also the behavior for many reporting tools that allow for offline report generation where-in the user triggering the report would download an empty CSV file if the report criteria is not satisfied.

Reply:
As discussed in the meeting, I will keep the behavior as is. Also, the presence or absence of data within the file does not really add much load to the system since the user anyways has to download the file to examine the data. and hence does not really save anything in terms of I/O.